Mathlib/Topology/Algebra/AsymptoticCone.lean

```diff
@@ -0,0 +1,380 @@
+  unfold asymptoticNhds
+  apply le_antisymm
+  · refine iSup_le fun u => ?_
+    simp_rw [vadd_eq_add, add_pure, ← map₂_smul, map_map₂, ← map_prod_eq_map₂]
+    have : (fun x : k × V => x.1 • x.2 + u) =ᶠ[atTop ×ˢ 𝓝 v]
+        (Function.uncurry (· • ·)) ∘ (fun x : k × V => (x.1, x.2 + x.1⁻¹ • u)) := by
+      filter_upwards [tendsto_fst.eventually (eventually_ne_atTop 0)] with _ h
+      simp [h]
+    rw [map_congr this, ← map_map]
+    apply map_mono
+    have : Tendsto (fun x : k × V => (x.1, x.2 + x.1⁻¹ • u)) (atTop ×ˢ 𝓝 v) _ :=
+      tendsto_fst.prodMk <| tendsto_snd.add <| tendsto_fst.inv_tendsto_atTop.smul_const u
+    simpa
+  · apply (le_iSup _ 0).trans'
+    simp
+
+theorem asymptoticNhds_eq_smul_vadd (v : V) (p : P) :
+    asymptoticNhds k P v = atTop (α := k) • 𝓝 v +ᵥ pure p := by
+  rw [← asymptoticNhds_eq_smul, asymptoticNhds_vadd_pure]
+
+instance {v : V} : (asymptoticNhds k P v).NeBot := by
+  have ⟨p⟩ : Nonempty P := inferInstance
+  rw [asymptoticNhds_eq_smul_vadd v p]
+  infer_instance
+
+private theorem asymptoticNhds_zero' : asymptoticNhds k V (0 : V) = ⊤ := by
+  rw [← top_le_iff, ← iSup_pure_eq_top, iSup_le_iff]
+  intro v
+  rw [← map_const (f := atTop (α := k))]
+  have : (fun _ => v) =ᶠ[atTop (α := k)]
+      (Function.uncurry (· • ·)) ∘ (fun c => (c, c⁻¹ • v)) := by
+    filter_upwards [eventually_ne_atTop 0] with _ h
+    simp [h]
+  rw [map_congr this, ← map_map, asymptoticNhds_eq_smul, ← map₂_smul, ← map_prod_eq_map₂]
+  apply map_mono
+  have : Tendsto (fun c => (c, c⁻¹ • v)) (atTop (α := k)) _ :=
+    tendsto_id.prodMk <| tendsto_inv_atTop_zero.smul_const v
+  simpa
+
+@[simp]
+theorem asymptoticNhds_zero : asymptoticNhds k P (0 : V) = ⊤ := by
+  have ⟨p⟩ : Nonempty P := inferInstance
+  rw [← asymptoticNhds_vadd_pure 0 p, asymptoticNhds_zero', vadd_pure]
+  exact (Equiv.vaddConst p).surjective.filter_map_top
+
+theorem _root_.Filter.Tendsto.atTop_smul_nhds_tendsto_asymptoticNhds {f : α → k} {g : α → V} {v : V}
+    (hf : Tendsto f l atTop) (hg : Tendsto g l (𝓝 v)) :
+    Tendsto (fun x => f x • g x) l (asymptoticNhds k V v) := by
+  rw [asymptoticNhds_eq_smul, ← map₂_smul, ← map_prod_eq_map₂]
+  exact tendsto_map.comp (hf.prodMk hg)
+
+theorem _root_.Filter.Tendsto.atTop_smul_const_tendsto_asymptoticNhds {f : α → k} (v : V)
+    (hf : Tendsto f l atTop) :
+    Tendsto (fun x => f x • v) l (asymptoticNhds k V v) :=
+  hf.atTop_smul_nhds_tendsto_asymptoticNhds tendsto_const_nhds
+
+theorem asymptoticNhds_smul (v : V) {c : k} (hc : 0 < c) :
+    asymptoticNhds k P (c • v) = asymptoticNhds k P v := by
+  have ⟨p⟩ : Nonempty P := inferInstance
+  simp_rw [asymptoticNhds_eq_smul_vadd _ p,
+    ← show map (c • ·) (𝓝 v) = 𝓝 (c • v) from
+      (Homeomorph.smulOfNeZero c hc.ne').map_nhds_eq v,
+    ← map₂_smul, map₂_map_right, smul_smul, ← map₂_map_left,
+    show map (· * c) atTop = atTop from (OrderIso.mulRight₀ _ hc).map_atTop]
+
+@[simp]
+theorem nhds_bind_asymptoticNhds (v : V) :
```

Comment:
I'm not sure if treating `bind` as an infix operator wrt the naming convention makes sense here